### ISSUE TITLE
Attention & Bidirectional LSTMs

### ISSUE THREAD

**Issue by kostasang** [slp-ntua/slp-labs#63]
Καλησπερα. Στην περίπτωση των bidirectional LSTMs εχουμε 2 ακολουθιες, μια forward και μια backward. Οι ακολουθιες ειναι ισες σε μηκος με με την padded ακολουθια εισοδου. Η ερωτηση ειναι η εξης:

Για να χρησιμοποιησουμε το μηχανισμο attention:

1) Τροφοδοτουμε για καθε χρονικη στιγμη σε ενα attention layer το concatenation forward|backward για καθε χρονικη στιγμη για αν προκύεψει τελικά το attention(forwards|backwards); 

2) Τροφοδοτουμε σε 1 attention layer πρωτα τη forward ακολουθια, ύστερα τη backward ακολουθία και τελικά παίρνουμε το attention(forward) | attention(backward);

3) Τροφοδοτουμε σε 2 διαφορετικά attention layers, ένα για τις Forward και ένα για τις backward ακολουθίες και τελικά παίρνουμε attention1(forward) | attention2(backward);

**Comment by pxaris** [comment]
Καλησπέρα. Το 1ο. 
Η μόνη διαφορά, σε σχέση με την υλοποίηση του απλού LSTM με attention, είναι ότι στο bidirectional LSTM τα `h_t` θα έχουν διπλάσιο μέγεθος (2 * hidden_size) και συνεπώς πρέπει να καθοριστεί κατάλληλα η διάσταση του Attention layer (attention_size). Αυτά τα `h_t` (κρυφή κατάσταση του biLSTM για κάθε βήμα `t` - forward || backward) θα δωθούν στο Attention layer.

**Comment by kostasang** [comment]
Το concatenation να γινει με τη σειρα που επιστρεφονται τα forward και backward?

Για το forward στη θεση 0 ειναι η εξοδος για το πρωτο timestep ενώ για το backward για το τελευταιο timestep (δηλαδή την τελευταία λέξη). Θα μου φαινοταν λογικο να κανω flip το backward ωστε καθε θεση να αναφερεται στην ιδια χρονικη στιγμη (στην ίδια λεξη), αλλά ετσι τα σκουπιδια λογω του padding θα βρισκονταν στο τελος για το forward και στην αρχη για το backward. Επομενως φανταζομαι το concatenation πρεπει να γινει με τη σειρα που επιστρεφονται.

**Comment by georgepar** [comment]
κανονικά τα flipαρεις, αγνοώντας το padding. Δηλαδή χρησιμοποιείς τα lengths που περνάς μέσα και φλιπάρεις μόνο από το 0 έως το length για κάθε sample του batch.

Τώρα η vectorized υλοποίηση σε αυτό είναι λίγο εξεζητημένη, αλλά ένα for loop είναι απλό.

**Comment by kostasang** [comment]
Επομενως πρεπει κάθε στοιχεία του concatenated sequence να αναφερεται στην ιδια λέξη, και για να γίνει αυτό κανω flip τα backward που με αφορουν δηλαδή μόνο οσα ΔΕΝ ειναι σκουπιδια, τα οποία παραμένουν στο τέλος μαζι με τα σκοπιδια του forward. 

Θα μποροουσατε να μας κατευθυνεται για τη vectorized υλοποιηση;

**Comment by pxaris** [comment]
Απ' όσο γνωρίζω, το PyTorch μέσω της κλάσης LSTM αντιστοιχίζει τα indexes του backward με εκείνα του forward αυτόματα. Συνεπώς, δεν χρειάζεται να κάνετε κάποια custom υλοποίηση γι αυτό. Δείτε μία [σχετική ερώτηση εδώ.](https://stackoverflow.com/questions/53010465/bidirectional-lstm-output-question-in-pytorch)

**Comment by kostasang** [comment]
"Yes, when using a BiLSTM the hidden states of the directions are just concatenated (the second part after the middle is the hidden state for feeding in the reversed sequence).
So splitting up in the middle works just fine."

Απ' οσο καταλαβαινω, το πρωτο στοιχειο της ακολουθιας εως τα μισα αναφερεται στην πρώτη λεξη της ακολουθιας εισόδου και απο τα μισα εως το τελος στην τελευταια λεξη της ακολουθίας εισόδου (γι αυτο και λεεί "the second part after the middle is the hidden state **for feeding in the reversed sequence**"). Επομένως το flipping του backward μου φαινεται λογικο.

**Comment by pxaris** [comment]
Sorry, my bad. Έχετε δίκιο

**Comment by kostasang** [comment]
Κανένα πρόβλημα!

**Comment by georgepar** [comment]
> Θα μποροουσατε να μας κατευθυνεται για τη vectorized υλοποιηση;

Όπως προείπα δεν είναι trivial, οπότε προτείνω καλύτερα να κάνετε τη λούπα.
Παρόλα αυτά παραθέτω κώδικα ο οποίος κάνει κάτι λίγο διαφορετικό.
Διαλέγει το τελευταίο hidden του forward και του backward και κάνει αυτά concatenate για να πάρει ένα representation για το τελευταίο token.

```python
    def _merge_bi(self, forward, backward):
        if self.merge_bi == 'sum':
            return forward + backward
        return torch.cat((forward, backward), dim=-1)

    def _select_last_unpadded(self, out, lengths):
        gather_dim = 1 if self.batch_first else 0
        gather_idx = ((lengths - 1)  # -1 to convert to indices
                      .unsqueeze(1)  # (B) -> (B, 1)
                      .expand((-1, self.hidden_size))  # (B, 1) -> (B, H)
                      # (B, 1, H) if batch_first else (1, B, H)
                      .unsqueeze(gather_dim))
        # Last forward for real length or seq (unpadded tokens)
        last_out = out.gather(gather_dim, gather_idx).squeeze(gather_dim)
        return last_out

    def _final_output(self, out, lengths):
        # Collect last hidden state
        # Code adapted from https://stackoverflow.com/a/50950188
        if not self.bidirectional:
            return self._select_last_unpadded(out, lengths)

        forward, backward = (out[..., :self.hidden_size],
                             out[..., self.hidden_size:])
        # Last backward corresponds to first token
        last_backward_out = (backward[:, 0, :]
                             if self.batch_first
                             else backward[0, ...])
        # Last forward for real length or seq (unpadded tokens)
        last_forward_out = self._select_last_unpadded(forward, lengths)
        return self._merge_bi(last_forward_out, last_backward_out)
```

**Comment by kostasang** [comment]
Ευχαριστω για τον κώδικα, ομοιαζει με αυτόν της αναγνωρισης προτυπων για την εξαγωγη των τελευταιων hidden states του forward & backward.

Μια τελευταια ερωτηση, σε περιπτωσεις που οι υλοποιησεις μας δεν ειναι vectorized και χρησιμοποιουμε for loops, το autograd του pytorch δεν εμφανιζει προβληματα με τη σχεδιαση των γραφου των operations που λαμβανουν μερος;

**Comment by georgepar** [comment]
Ναι θα γίνει σωστά.

**Comment by kostasang** [comment]
Σορι και πάλι, αλλά απ' οτι βλεπω το τελευταιο time step του backward (δηλαδή το backward output για την πρωτη λεξη του input sequence) λαμβάνεται με τον εξής τρόπο **backward[:, 0, :]**. Επιπλέον, το forward της πρώτης λέξης είναι παλι forward[:,0,:]. 

Επομενως για την πρώτη λεξη του input sequence έχουμε forward[:,0,:] και backward[:,0,:]. Αρα κανενα flipping δεν χρειαζεται. Ο @pxaris ειχε δικιο τελικα;

**Comment by pxaris** [comment]
Κοιτώντας [το issue εδώ](https://github.com/pytorch/pytorch/issues/3587#issuecomment-360168707), βλέπω ότι υπάρχει διαφορετική αντιμετώπιση όσον αφορά το `output` και το `hidden state`. Το `output` του κάθε βήματος ρυθμίζεται κατάλληλα από το framework ώστε να αντιστοιχεί στο ίδιο στοιχείο `x_i` της ακολουθίας τόσο για το forward όσο και για το backward LSTM.
Δηλαδή με `output, hn = lstm([x_1, x_2, x_3, x_4, x_5])`, το `output` θα είναι:
```
[F(x_1), B(x_1)]
[F(x_2), B(x_2)]
[F(x_3), B(x_3)]
[F(x_4), B(x_4)]
[F(x_5), B(x_5)]
```
το οποίο είναι αυτό που επιθυμούμε (κάτι που δεν ισχύει για το hidden state). 
Συνεπώς, από τη στιγμή που εμείς χρειαζόμαστε το output (`h_t`) για κάθε `t` χρησιμοποιούμε κατευθείαν αυτό που επιστρέφει το PyTorch.

**Comment by kostasang** [comment]
Τελεια, ευχαριστω και σορι για το μπερδεμα πριν!
